src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs

```
@@ -923,16 +925,29 @@ public override Task StartAsync()
         }
 
         /// <inheritdoc />
-        public override async Task<string> UpdateStateAsync(
+        public override Task<string> UpdateStateAsync(
             OrchestrationRuntimeState newRuntimeState,
             OrchestrationRuntimeState oldRuntimeState,
             string instanceId,
             string executionId,
             string eTagValue)
+        {
+            throw new NotSupportedException("Use the overload that includes trackingStoreData");
+        }
+
+        /// <inheritdoc />
+        public async Task<string> UpdateStateAsync(
+            OrchestrationRuntimeState newRuntimeState,
+            OrchestrationRuntimeState oldRuntimeState,
+            string instanceId,
+            string executionId,
+            string eTagValue,
+            object trackingStoreData)
         {
             int estimatedBytes = 0;
             IList<HistoryEvent> newEvents = newRuntimeState.NewEvents;
             IList<HistoryEvent> allEvents = newRuntimeState.Events;
+            TrackingStoreData data = (TrackingStoreData) trackingStoreData;
```

Why can't the parameter type be `TrackingStoreData`? That way, we'd skip the casting in line 950.
I would prefer the callers cast their arguments at the callsites

---

The mechanism is supposed to be independent of the implementation of TrackingStore. So it can't really depend on a type specifically designed for AzureTableTrackingStore.

---

In that case, should we do a check here in case the type casting fails?

---

I am not sure what benefit an explicit check would add over the exception that is already being thrown if this should ever happen. There is not much risk here that the cast would ever fail IMO.

---

Now that I think about it, you're right - the cast should fail if it's not valid. I was getting confused with the behavior of casting in PLs with gradual typing, where sometimes the cast fails silently. Ok then, please disregard my comment here.